File: backend/fastapi/app/services/templates.py
import re
from jinja2 import Environment, FileSystemLoader, select_autoescape
from markupsafe import Markup
from pathlib import Path

TEMPLATE_DIR = Path(__file__).resolve().parent.parent / "templates"
env = Environment(loader=FileSystemLoader(str(TEMPLATE_DIR)), autoescape=select_autoescape(['html','xml']))

def nl2br(value: str) -> str:
    """Convert newlines in ``value`` to HTML ``<br>`` tags."""
    return Markup(value.replace('\n', '<br>\n'))

env.filters['nl2br'] = nl2br

def render_letter(template_name: str, context: dict) -> str:
    """Render a letter template with the provided ``context``."""
    return env.get_template(template_name).render(**context)

File: backend/express/src/services/templates.ts
import nunjucks from 'nunjucks';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const TEMPLATE_DIR = path.resolve(__dirname, '../../../fastapi/app/templates');
const env = nunjucks.configure(TEMPLATE_DIR, { autoescape: true });

env.addFilter('nl2br', (str: string) => str.replace(/\n/g, '<br>\n'));

/** Render a letter template with the provided context. */
export function renderLetter(template: string, context: Record<string, any>): string {
  return env.render(template, context);
}
